### PR TITLE
saving progset without wiping out existing programs

### PR DESCRIPTION
**Pull request by Ashalynd**
https://trello.com/c/uapmGuzu/824-when-progset-is-saved-all-existing-programs-are-recreated-from-scratch

@nanuxbe please review

### CODE REVIEW

File: server/webapp/dbmodels.py [hunk 1]
```diff
@@ -660,32 +660,51 @@ def restore(self, progset, program_list):
...
-    def create_programs_from_list(self, programs):
+    def recreate_programs_from_list(self, programs, progset_id):
         from optima.utils import saves
 
         prog_shorts = []
-        for program in programs:
-            kwargs = {}
-            for field in ['name', 'short', 'category', 'targetpops', 'pars', 'costcov', 'criteria']:
-                kwargs[field] = program[field]
+        desired_shorts = set([program.get('short', program.get('short_name', '')) for program in programs])
```
Comment (nanuxbe) [inline server/webapp/dbmodels.py]: should be the other way around, short_name is the default with fallback to short

File: server/webapp/dbmodels.py [hunk 2]
```diff
@@ -660,32 +660,51 @@ def restore(self, progset, program_list):
...
+                existing_shorts[program.short]=program
+        db.session.flush()
 
-            if kwargs['short'] in prog_shorts:
-                raise DuplicateProgram(kwargs['short'])
+
+        for program in programs:
+            # Kind of a hack but sometimes we receive short ans sometimes short_name
+            short = program.get('short', program.get('short_name', ''))
+            if short in existing_shorts:
```
Comment (nanuxbe) [inline server/webapp/dbmodels.py]: same here

File: server/webapp/dbmodels.py [hunk 3]
```diff
@@ -660,32 +660,51 @@ def restore(self, progset, program_list):
...
-        for program in programs:
-            kwargs = {}
-            for field in ['name', 'short', 'category', 'targetpops', 'pars', 'costcov', 'criteria']:
-                kwargs[field] = program[field]
+        desired_shorts = set([program.get('short', program.get('short_name', '')) for program in programs])
+        existing_programs = db.session.query(ProgramsDb).filter_by(progset_id=progset_id)
 
-            # Kind of a hack but sometimes we receive short ans sometimes short_name
-            if 'short_name' in program and program['short_name'] is not None:
-                kwargs['short'] = program['short_name']
```
Comment (nanuxbe) [inline server/webapp/dbmodels.py]: see in the old code here. short is used, but then if short_name exists, short is replaced by short_name

**Comment by robynstuart** [comment]
thanks @Ashalynd !!

**Pull request merged**